effect/README.md

```diff
@@ -0,0 +1,294 @@
+
+Use `IO` because it's simply not practical to write real-world, correct software without it.
+
+# Usage
+
+## Main
+
+Your main function can extend `SafeApp`, which provides a complete runtime
+system and allows your entire program to be purely functional.
+
+```scala
+import scalaz.effect.{IO, SafeApp}
+import scalaz.effect.console._
+
+object MyApp extends SafeApp {
+  def run(args: IList[String]): IO[Unit] =
+    for {
+      _ <- putStrLn("Hello! What is your name?")
+      n <- getStrLn
+      _ <- putStrLn("Hello, " + n + ", good to meet you!")
+    } yield ()
+}
+```
+
+## Pure Values
+
+You can lift pure values into `IO` with `IO.point`:
+
+```scala
+val liftedString: IO[String] = IO.point("Hello World")
+```
+
+The constructor uses non-strict evaluation, so the parameter will not be evaluated until when and if the `IO` action is executed at runtime.
+
+Alternately, you can use the `IO.now` constructor for strict evaluation:
+
+```scala
+val lifted: IO[String] = IO.now("Hello World")
+```
+
+You should never use either constructor for impure code. The result of doing so is undefined, and will probably result in runtime errors.
+
+## Impure Code
+
+You can use the `sync` method of `IO` to import effectful synchronous code into your purely functional program:
+
+```scala
+def readFile(f: String): IO[ByteArray] = IO.sync(Bytes.readFile(f))
+```
+
+You can use the `async` method of `IO` to import effectful asynchronous code into your purely functional program:
+
+```scala
+def makeRequest(req: Request): IO[Response] =
+  IO.async(cb => Http.req(req, cb))
+```
+
+## Mapping
+
+You can change an `IO[A]` to an `IO[B]` by calling the `map` method with a function `A => B`. This lets you transform values produced by actions into other values.
+
+```scala
+val answer = IO.point(21).map(_ * 2)
+```
+
+## Chaining
+
+You can execute two actions in sequence with the `flatMap` method. The second action may depend on the value produced by the first action.
+
+```scala
+val contacts: IO[IList[Contact]] =
+  readFile("contacts.csv").flatMap((file: ByteArray) =>
+    parseCsv(file).map((csv: IList[CsvRow]) =>
+      csv.map(rowToContact)
+    )
+  )
+```
+
+You can use Scala's `for` comprehension syntax to make this type of code more compact:
+
+```scala
+val contacts: IO[IList[Contact]] =
+  for {
+    file <- readFile("contacts.csv")
+    csv  <- parseCsv(file)
+  } yield csv.map(rowToContact)
+```
+
+## Failure
+
+You can create `IO` actions that describe failure with `IO.fail`:
+
+```scala
+val io: IO[String] = IO.fail(new Error("Oh noes!"))
+```
+
+Like all `IO` values, these are immutable values and do not actually throw any exceptions; they merely describe failure as a first-class value.
+
+You can surface failures with `attempt`, which takes an `IO[A]` and produces an `IO[Throwable \/ A]`:
+
+```scala
+val result = openFile("data.json").attempt.map {
```

Comment:
Where is this awesome-sauce? I've been curious about how one could have truly non-blocking IO, which called back to the rts rather than having to waste a thread with something like `Files.readAllBytes(...)`.  Java 8 has some sort of stuff in java.nio, Selector, SelectableChannel, e,g. that seems to offer some non-blocking facilities that look quite painful to manage as a client and perfect for rolling into something like this so clients can just program naturally without worrying about the nitty gritty of selecting, registering keys, reading more bytes when ready, etc.  When I first saw this readme a few days ago I was hopeful that truly non-blocking yet sensibly concurrent/parallel file operations existed, but I guess they're just a concept/WIP?  If not please point me to where I can look at how they're hooked up!

Reply:
It does not exist yet, but will most definitely exist soon. 😄 

The first step is `scalaz-nio`.

Reply:
Ok, so I played with how one might read a file (hopefully some file manipulation can be rolled into the library), but I can't quite get it the final mile (i.e. so that each chunk read is done from the RTS so that cancellation or fiber-timeslicing can occur) - sorry, but I just can't my head around how to convert impure code i.e. where making the same call could have different results (bytes read) into pure IO/effect code and how to keep making that call like a do while in the IO RTS…  Any advice/insight appreciated!

https://gist.github.com/scr/9d9fd533f02f7dcf4c4029ab55901412

Reply:
I see there's a `forever` and `retry` but both of those work on failures - it would seem that a tri-state is needed, or `Throwable \/ Option[T]` with something like a `.doWhile` or `.doUntil` operator that calls the code until it receives either the throwable or an empty/None result (or nonEmpty/Some result if doUntil.

Reply:
or some sort of `IO.trampoline` thing?

Reply:
Ok… got it working with existing primitives (thanks @jdegoes for nudge to try IO.flatten); updated gist to show how IO.flatten of the IO.async[IO[A]] lets you chain File work indefinitely to read the entire file in and give chances for cooperation with this Fiber stuff!  So cool!